# Модуль2

In [1]:
!pip install xgboost

In [2]:
import pandas as pd
import numpy as np
import joblib 
import matplotlib.pyplot as plt
import seaborn as sns
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import precision_recall_curve, classification_report
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score

In [3]:
train_df=pd.read_csv("data/full_train_df.csv", index_col='session_id')

In [4]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 253561 entries, 1 to 253561
Data columns (total 14 columns):
 #   Column   Non-Null Count   Dtype
---  ------   --------------   -----
 0   site1    253561 non-null  int64
 1   site2    253561 non-null  int64
 2   site3    253561 non-null  int64
 3   site4    253561 non-null  int64
 4   site5    253561 non-null  int64
 5   site6    253561 non-null  int64
 6   site7    253561 non-null  int64
 7   site8    253561 non-null  int64
 8   site9    253561 non-null  int64
 9   site10   253561 non-null  int64
 10  morning  253561 non-null  int64
 11  day      253561 non-null  int64
 12  evening  253561 non-null  int64
 13  target   253561 non-null  int64
dtypes: int64(14)
memory usage: 29.0 MB


In [5]:
Y=train_df['target']
X=train_df.drop('target', axis=1)

Разделим выборку на обучающую и тестовую в соотношении 70/30. Выборка имеет 253561 наблюдения, это достаточно много и позволит оставить для тестирования 30%.

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, Y,   test_size=0.3, random_state=42)

В качестве метрики выберем AUC-ROC. Так как в данной задаче имеем дело с несбалансированной выборкой. Критерий AUC-ROC устойчив к несбалансированным классам 

In [7]:
lr = LogisticRegression(random_state=42)
lr.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [8]:
y_pred_lr = lr.predict(X_test)

In [9]:
roc_auc_score(y_test, y_pred_lr)

0.5

In [10]:
RF = RandomForestClassifier(random_state=42)
RF.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [11]:
y_pred_rf = RF.predict(X_test)

In [12]:
roc_auc_score(y_test, y_pred_rf)

0.5176984239665022

In [13]:
GB=XGBClassifier(random_state=42)
GB.fit(X_train, y_train)

/opt/anaconda3/lib/python3.8/site-packages/xgboost/sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[11:44:03] WARNING: /Users/travis/build/dmlc/xgboost/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=12, num_parallel_tree=1, random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [14]:
y_pred_GB = GB.predict(X_test)

In [15]:
roc_auc_score(y_test, y_pred_GB)

0.6011257715684213

XGBoost показал лучшее какчество, чем логистическая регрессия и случайный лес